### ISSUE TITLE
Non-exported public type in return type of pub fn in exported public type

### ISSUE THREAD

**Issue by caranatar** [rust-lang/rust#75868]
I ran into a weird issue and I'm not sure if it's a compiler bug, a rustdoc bug, or neither. I took over a crate which has a public struct (`Rule`), inside a non-public module, which does not get exported. That type gets used in the return type of a public function (`get_rule`) in an exported public type (`Grammar`), [docs.rs here](https://docs.rs/tracery/0.1.3/tracery/struct.Grammar.html#method.get_rule). I created what I think is a minimal reproduction of the issue

[playground link here](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018)
```rust
pub mod c {
    mod a {
        pub struct Foo;
    }

    mod b {
        use crate::c::a::Foo;
        pub fn bar() -> Foo {
            Foo {}
        }
    }

    pub use b::bar;
}

fn main() {
    let foo = c::bar();
    // fails as expected:
    //let foo = c::a::Foo {};
}
```

I would expect this to be flagged as leaking a private type in a public function, but it isn't on stable or nightly. And if it shouldn't be flagged, then I'd expect rustdoc to generate documentation for the type, but as you can see in the docs.rs link above, it doesn't. Is one of them doing the wrong thing?

**Comment by jonas-schievink** [comment]
This is working as expected. The private-in-public lint only looks at the items visibility modifier. I think there's an accepted RFC to change or clarify this behavior. And rustdoc has no way to document the struct since it only documents the externally usable portion of the API (unless you pass `--document-private-items`).
